### PR TITLE
Fix batched LU factorization grain_size heuristic for large matrices (#181570)

### PR DESCRIPTION
**Pull request by agramfort**
Summary:

The parallelization heuristic in `apply_lu_factor` computes `chunk_size_per_thread` as `min(1.0, 3200 / rank^3)` and then `grain_size = chunk_size_per_thread * num_threads`. For large matrices (rank > ~14), this truncates to `chunk_size_per_thread = 0`, producing `grain_size = 0`. However, `at::parallel_for` interprets `grain_size=0` as "always parallelize" (since `numiter > 0` is trivially true), which is the opposite of the intended behavior.

This causes multiple threads to call `dgetrf_` concurrently, triggering MKL thread-safety issues in `DLASWP` (row pivot swap). The symptom is `Intel oneMKL ERROR: Parameter 6 was incorrect on entry to DLASWP` errors and incorrect numerical results when using `torch.linalg.inv` on batched large matrices with `num_threads > 1`.

The fix: when `chunk_size_per_thread` truncates to 0, set `grain_size = batch_size` so the entire batch runs on a single thread, which was the original intent of the heuristic for large matrices.

Test Plan:
Added `test_inverse_batched_multithread` which runs batched `linalg.inv` on 256x256 matrices with 4 threads. Before this fix, this would produce MKL errors and incorrect results. After the fix, the residual `||A @ A_inv - I||` is within tolerance.

buck2 test //caffe2/test:linalg -- -r test_inverse_batched_multithread

Differential Revision: D102573456

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

File: test/test_linalg.py [hunk 1]
```diff
@@ -10551,6 +10551,35 @@ def test_logaddexp_cpu_vs_cuda_complex(self, device):
...
+        # threads used to trigger MKL DLASWP thread-safety errors because
+        # apply_lu_factor's grain_size heuristic produced grain_size=0 for
+        # large matrices, which at::parallel_for interprets as "always
+        # parallelize" instead of the intended "serialize".
+        num_threads = torch.get_num_threads()
+        try:
+            torch.set_num_threads(4)
+            n = 256
+            batch = 8
+            A = torch.randn(batch, n, n, dtype=dtype, device=device)
```
Comment (lezcano) [inline test/test_linalg.py:11472]: how is this related to the rank 14 discussed in the OP?

**Comment by agramfort** [comment]
to replicate the bug you can run

`python script.py --num_threads=2 --dim=300 --n-samples=200`

where script.py is:

```
import argparse

import numpy as np
import torch


def main(num_threads: int, dim: int, n_samples: int) -> None:
    rng = np.random.RandomState(42)
    A = torch.from_numpy(rng.randn(n_samples, dim, dim))

    torch.set_num_threads(num_threads)
    A_inv = torch.linalg.inv(A)
    res = float(torch.linalg.norm((A @ A_inv - torch.eye(dim).unsqueeze(0)).flatten()))
    print(f"{res=}")


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--num_threads", type=int, default=1)
    parser.add_argument("--dim", type=int, default=300)
    parser.add_argument("--n-samples", type=int, default=200)
    args = parser.parse_args()
    main(num_threads=args.num_threads, dim=args.dim, n_samples=args.n_samples)
```

File: test/test_linalg.py (hunk 1 above)
Comment (agramfort) [reply test/test_linalg.py:11472]: `n` just needs to be large enough to trigger the crash. 14 is just because of the 3200 / 14**3 where 14 is rank in the code.

**Review by lezcano — APPROVED** [review]
AH, rank as the dimension of the ambient space, not as in the number of dimensions of the operator. Sure

**Review by nikitaved — APPROVED** [review]
